cli/analyze/analyze.go

```diff
@@ -0,0 +1,559 @@
+		for _, input := range r.RuleInput {
+			// Don't index inputs from external repos.
+			if strings.HasPrefix(input, "@") {
+				continue
+			}
+			graph.DirectDependents[input] = append(graph.DirectDependents[input], r)
+		}
+	}
+	return graph, nil
+}
+
+// TransitiveDependentsCount returns the number of transitive dependents of a
+// given target, counting the target itself.
+func (g *DependencyGraph) TransitiveDependentCount(name string) int {
+	// Traverse the graph of transitive dependents, keeping track of nodes
+	// already visited.
+
+	// TODO: avoid a full graph traversal on each function call.
+	frontier := []string{name}
+	visited := map[string]bool{}
+	for len(frontier) > 0 {
+		name := frontier[len(frontier)-1]
+		frontier = frontier[:len(frontier)-1]
+		if visited[name] {
+			continue
+		}
+		visited[name] = true
+		for _, r := range g.DirectDependents[name] {
+			frontier = append(frontier, *r.Name)
+		}
+	}
+
+	return len(visited)
+}
+
+// LongestPath returns the longest dependency path from one node in the graph to
+// any other node.
+func (g *DependencyGraph) LongestPath() []string {
+	longestPathLength := 0
+	var longestPathEnd *string
+
+	length := map[string]int{}
+	pred := map[string]*string{}
+	e := g.EdgeSet()
+	for _, m := range g.TopologicalSort() {
+		for n := range e.Incoming[m] {
+			l := length[n] + 1
+			if l > length[m] {
+				length[m] = l
+				n := n
+				pred[m] = &n
+			}
+			if l > longestPathLength {
+				longestPathLength = l
+				m := m
+				longestPathEnd = &m
+			}
+		}
+	}
+	if longestPathEnd == nil {
+		return nil
+	}
+	path := []string{*longestPathEnd}
+	for {
+		p := pred[path[len(path)-1]]
+		if p == nil {
+			break
+		}
+		path = append(path, *p)
+	}
+	reverseSlice(path)
+	return path
+}
+
+// Roots returns all labels in the graph which have no direct dependents.
+func (g *DependencyGraph) Roots() []string {
+	var roots []string
+	for name := range g.Rules {
+		if len(g.DirectDependents[name]) == 0 {
+			roots = append(roots, name)
+		}
+	}
+	return roots
+}
+
+func (g *DependencyGraph) EdgeSet() *EdgeSet {
+	outgoing := map[string]map[string]bool{}
+	incoming := map[string]map[string]bool{}
+	s := g.Roots()
+	for len(s) > 0 {
+		n := s[len(s)-1]
+		s = s[:len(s)-1]
+		if _, ok := outgoing[n]; ok {
+			continue // already visited
+		}
+		rule := g.Rules[n]
+		if rule == nil {
+			continue
+		}
+		for _, m := range rule.RuleInput {
+			if outgoing[n] == nil {
+				outgoing[n] = map[string]bool{}
+			}
+			if incoming[m] == nil {
+				incoming[m] = map[string]bool{}
+			}
+			outgoing[n][m] = true
+			incoming[m][n] = true
+
+			s = append(s, m)
+		}
+	}
+	return &EdgeSet{Outgoing: outgoing, Incoming: incoming}
+}
+
+// TopologicalSort returns all node labels in the graph sorted so that each node
+// is sorted before any of its transitive deps.
+func (g *DependencyGraph) TopologicalSort() []string {
+	// https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm
```

Comment:
Not to be confused with [Kuhn's algorithm](https://en.wikipedia.org/wiki/Kuhn%27s_algorithm).